pulp_container/app/registry_api.py

```diff
@@ -204,7 +205,7 @@ def authentication_classes(self):
         List of authentication classes to check for this view.
         """
         if settings.get("TOKEN_AUTH_DISABLED", False):
-            return [RegistryAuthentication]
+            return [RemoteUserRegistryAuthentication, RegistryAuthentication]
```

Comment:
can you please remind me whether the order matters and if yes how this is evaluated?

Reply:
The order matters. The authentication procedure ends when one of the authentication classes succeeds and returns an authenticated user from its method `authenticate`. If none of the authentication classes succeeds, a user shall not be authenticated.

There might be a problem with the Nginx configuration where data from both the basic authentication and remote authentication methods are passed to Pulp at the same time. Pulp will not be able to determine which method should be used and traverses the list of authentication classes one by one. Do you think we should check the Pulp settings and set the list of authentication classes accordingly to prevent unnecessary failures that could not be seen after misconfiguring Nginx?

```python
if  "pulpcore.app.authentication.PulpRemoteUserAuthentication" in settings.REST_FRAMEWORK["DEFAULT_AUTHENTICATION_CLASSES"]:
    return [RemoteUserRegistryAuthentication]  # remote auth
else:
    return [RegistryAuthentication]  # basic auth
```

Reply:
looking at our docs we mention how to configure remote user auth and that seems to remove basic auth but keep session auth. In that case the webserver would not have such dilemma. I am not sure whether a configuration where both remote user and basic auth  enabled is valid? @bmbouter can you comment?

Reply:
The thinking in those docs (I wrote years ago) is that when a user opts to use webserver auth they want everyone to auth with that and there shouldn't be another "side-door" of users so to speak. Removing the basic auth removes that "side-door" authentication mechanism basically.

Perhaps one use case (a guess really) would be if an admin setup Pulp and used the webserver auth to connect all of their enterprise users to it, but then they wanted a few built-in users as admins to debug it if the enterprise auth was misconfigured or unavailable.

Is this helpful at all?